### PR TITLE
[RFC] Run all unit tests in separate processes

### PR DESCRIPTION
**Pull request by ZyX-I**
Should prevent all possible issues with state inherited from previous tests. Should also allow test to crash and still proceed with other tests.

Currently tests are also run with garbage collection disabled.

Note: tests are run in separate process. `before_each`, `after_each`, module code, etc is run in main process. So if you have conflicting imports or something like this this PR is not going to help.

Part of #5119.

After PR is merged it should also be possible to enable the same system which checks for core dumps which functional tests currently use, but I did not do that in this PR, primary because I do not need this much: except for rare occasions like with `tv_list_join()` unit tests normally either do not crash both locally and on CI or crash both locally and on CI.

### CODE REVIEW

**Comment by justinmk** [comment]
macOS has some core dump: https://s3.amazonaws.com/archive.travis-ci.org/jobs/207806481/log.txt

**Comment by justinmk** [comment]
After this change what happens if someone accidentally uses `it()` in a unit test? Should we make it an error?

**Comment by ZyX-I** [comment]
> macOS has some core dump: https://s3.amazonaws.com/archive.travis-ci.org/jobs/207806481/log.txt

I saw, and there are way too much errors. Was going to try some Mac OS image on virtualbox, though one found on the [internet](https://techsviewer.com/how-to-install-mac-os-x-el-capitan-on-pc-on-virtualbox/) already fails with blank screen and virtualbox critical error.

> After this change what happens if someone accidentally uses it() in a unit test? Should we make it a testlint error?

Not sure how error may be done. Though nothing particularly bad is going to happen, just state may leak and affect other tests.

**Comment by ZyX-I** [comment]
So far any attempt to run Mac OS in virtual machine results in that machine crashing, checked with different images and virtualbox or vmware.

**Comment by ZyX-I** [comment]
@jamessan Any ideas why 9aa0132 is needed for test to pass?

**Comment by ZyX-I** [comment]
Got that working on travis OS X. And not working in other places: `lfs.attributes` appears to return `nil` randomly, like if somebody was cleaning up when it is running (should not be possible as long as main process uses `sc.wait(child_pid)`).

**Comment by ZyX-I** [comment]
Tests passed, should be ready.

File: test/unit/tempfile_spec.lua [hunk 1]
```diff
@@ -1,58 +1,64 @@
...
 describe('tempfile related functions', function()
-  before_each(function()
-    tempfile.vim_deltempdir()
-  end)
-  after_each(function()
-    tempfile.vim_deltempdir()
+  before_each(deferred_call(function()
+    lib.vim_deltempdir()
+  end))
+  separate_cleanup(function()
```
Comment (justinmk) [inline test/unit/tempfile_spec.lua]: What is `separate_cleanup`? How is it different from `teardown` and `after_each`? A brief comment attached to `separate_cleanup` would help. Probably `deferred_call` also.

Comment (ZyX-I) [inline test/unit/tempfile_spec.lua]: `separate_cleanup` is run inside a new (child) process, as well as `deferred_call`. I think that actually code needs some altering.

**Comment by justinmk** [comment]
Test lint:

```
make[4]: Entering directory `/home/travis/build/neovim/neovim/build'
Checking /home/travis/build/neovim/neovim/test/unit/helpers.lua [0m[31m[1m4 warnings[0m

    /home/travis/build/neovim/neovim/test/unit/helpers.lua:172:5: setting non-standard global variable [0m[1mcdefs[0m
    /home/travis/build/neovim/neovim/test/unit/helpers.lua:176:5: setting non-standard global variable [0m[1mcdefs[0m
    /home/travis/build/neovim/neovim/test/unit/helpers.lua:214:22: accessing undefined variable [0m[1mcdefs[0m
    /home/travis/build/neovim/neovim/test/unit/helpers.lua:215:7: accessing undefined variable [0m[1mcdefs[0m
```

**Pull request merged**